### PR TITLE
Allow users to connect to the server

### PR DESCRIPTION
**Pull request by piecubed**

### CODE REVIEW

File: src/server/server_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+func StartServer() (*utils.WaitGroupCounter, *http.Server, *ServerHub) {
+	wg := utils.NewWaitGroupCounter()
+	srv, hub := Serve("localhost:8080", wg)
+	wg.Wait()
+	return wg, srv, hub
+}
+
+func ConnectToServer(t *testing.T) *websocket.Conn {
```
Comment (JosiahWI) [inline src/server/server_test.go:21]: Test makes no assertion. Could maybe assert that the error is nil.

File: src/server/server_test.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	u := url.URL{Scheme: "ws", Host: "localhost:8080", Path: "/ws"}
+	c, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
+	if err != nil {
+		t.Fatalf("Error connecting to localhost:8080: %v", err)
+	}
+	return c
+}
+
+func TestEnsureAllGoroutinesStopWhenClientExits(t *testing.T) {
```
Comment (JosiahWI) [inline src/server/server_test.go:30]: Pretty lengthy test, makes no assertion. Could maybe assert that start != end.

File: src/server/server_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	hub.mainLoopEvent <- &QuitMainLoop{}
+	err := srv.Shutdown(context.TODO())
+	if err != nil {
+		t.Fatalf("Error shutting server down: %v", err)
+	}
+}
+
+func TestEnsureAllGoroutinesStopWhenServerExits(t *testing.T) {
```
Comment (JosiahWI) [inline src/server/server_test.go:65]: Makes no assertion. Could maybe assert that `wg.Counter == 0`.

**Review by JosiahWI — COMMENTED** [review]
Tests don't seem to show the flow of data through the system. In fact it's not clear from reading these tests that the server does anything at all except start and stop.

File: src/server/server_test.go (hunk 1 above)
Comment (piecubed) [reply src/server/server_test.go:21]: This is just the setup for tests.  Tests are shown by `TestXxx`

Go does not have `assert`.  In testing, you use `if <cond> ... t.Fatalf("Uh oh")`

You may see the following code:
```
	if err != nil {
		t.Fatalf("Error connecting to localhost:8080: %v", err)
	}
```

I assert that err is not nil and error if it is.

File: src/server/server_test.go (hunk 2 above)
Comment (piecubed) [reply src/server/server_test.go:30]: > Could maybe assert that start != end.

```
if start != end {
		t.Log("failed, goroutines still running:")
		t.Logf("Start: %v, End: %v", start, end)
		wg.Names.Range(
			func(key, value any) bool {
				t.Log(key)
				return true
			},
		)
		t.FailNow()
	}
	
```

> Pretty lengthy test

The test tests the actual running state of the server, so in order to set the server up correctly it is indeed lengthy.  I could take out some of the code however into a helper function.

File: src/server/server_test.go (hunk 3 above)
Comment (piecubed) [reply src/server/server_test.go:65]: You may note the following code:
```
	if wg.Counter != 0 {
		t.Log("failed, goroutines still running:")
		wg.Names.Range(
			func(key, value any) bool {
				t.Log(key)
				return true
			},
		)
		t.FailNow()
	}
```

I appear to check if `wg.Counter != 0` and fail if it is.

**Comment by piecubed** [comment]
> Tests don't seem to show the flow of data through the system. In fact it's not clear from reading these tests that the server does anything at all except start and stop.

This PR is still a work in progress, and is not ready for review.  I simply made the PR to track my changes and show that this was actively being worked on.

The server does, indeed only start and stop.  The tests are to ensure starting and stopping works correctly for future tests, so there's not runaway goroutines impacting other tests.

**Comment by piecubed** [comment]
Closing this, as the scope has changed. See #16

**Pull request closed without merging**